File: Projects/codechef/VACCINE1.java
package codechef;

import java.util.Scanner;

class VACCINE1 {
    public static void main(String[] args) throws Exception {
        Scanner stdin = new Scanner(System.in);
        int D1 = stdin.nextInt();
        int v1 = stdin.nextInt();
        int D2 = stdin.nextInt();
        int v2 = stdin.nextInt();
        int p = stdin.nextInt();
        int countVaccine = 0;
        int count = 0;
        while (countVaccine < p) {
            if (count >= D1) {
                countVaccine += v1;
            }
            if (count >= D2) {
                countVaccine += v2;
            }
            count++;
        }
        System.out.println(count - 1);
    }
}

File: Projects/codechef/VACCINE1.kt
package codechef

import java.io.IOException
import java.util.Scanner

data class Company(val d: Int, val v: Int)

fun vaccine1(d1: Int, v1: Int, d2: Int, v2: Int, p: Int): Int {
    val companies = mutableListOf(Company(d1, v1), Company(d2, v2))
    companies.sortBy { it.d }
    companies.add(Company(Int.MAX_VALUE, 0))
    var v = 0
    var deficit = p
    for ((curr, next) in companies.zipWithNext()) {
        v += curr.v
        if (curr.d == next.d) continue
        val delta = if (next.d == Int.MAX_VALUE) Int.MAX_VALUE else (next.d - curr.d) * v
        if (deficit > delta) {
            deficit -= delta
        } else {
            return curr.d + (deficit - 1) / v
        }
    }
    return -1
}

fun main(args: Array<String>) {
    val stdin = Scanner(System.`in`)
    val d1 = stdin.nextInt()
    val v1 = stdin.nextInt()
    val d2 = stdin.nextInt()
    val v2 = stdin.nextInt()
    val p = stdin.nextInt()
    val d = vaccine1(d1, v1, d2, v2, p)
    println(d)
    stdin.close()
}
